File: 2001-2500/2192_getAncestors.py
class Solution:
    def getAncestors(self, n: int, edges: List[List[int]]) -> List[List[int]]:
        graph = {i:[] for i in range(n)}
        in_degree = {i:0 for i in range(n)}

        for start, end in edges:
            graph[start].append(end)
            in_degree[end] += 1

        queue = deque([node for node, degree in in_degree.items() if not degree])
        answer = [set() for _ in range(n)]

        while queue:
            node = queue.popleft()
            
            for neighbor in graph[node]:
                answer[neighbor].add(node)
                
                for element in answer[node]:
                    answer[neighbor].add(element)

                in_degree[neighbor] -= 1

                if in_degree[neighbor] == 0:
                    queue.append(neighbor)

        return [sorted(list(ans)) for ans in answer]

# Alternative solution

class Solution:
    def getAncestors(self, n: int, edges: List[List[int]]) -> List[List[int]]:
        graph = {i : [] for i in range(n)}
        outdegree = {i : 0 for i in range(n)}
        ancestors = [set() for i in range(n)]
        
        for src, des in edges:
            graph[des].append(src)
            outdegree[src] += 1
        
        start = [key for key, val in outdegree.items() if val == 0]
        visited = {key for key, val in graph.items() if len(val) == 0}

        def dfs(node):
            for ancestor in graph[node]:
                if ancestor not in visited:
                    visited.add(node)
                    dfs(ancestor)
                ancestors[node].add(ancestor)
                ancestors[node] |= ancestors[ancestor]

        for node in start:
            dfs(node)

        return [sorted(list(l)) for l in ancestors]

File: 2001-2500/2192_getAncestors.java
class Solution {
    public List<List<Integer>> getAncestors(int n, int[][] edges) {
        ArrayList<Integer>[] graph = new ArrayList[n];
        int[] indegree = new int[n];
        List<HashSet<Integer>> ancestors = new ArrayList<>();
        Queue<Integer> queue = new LinkedList<>();

        for (int i = 0; i < n; i++) {
            graph[i] = new ArrayList<>();
            ancestors.add(new HashSet<>());
        }

        for (int[] edge: edges) {
            int from = edge[0];
            int to = edge[1];

            graph[from].add(to);
            indegree[to]++;
        }

        for (int i = 0; i < n; i++) {
            if (indegree[i] == 0) {
                queue.add(i);
            }
        }

        while (!queue.isEmpty()) {
            int length = queue.size();

            for (int i = 0; i < length; i++) {
                int node = queue.remove();

                for (int child: graph[node]) {
                    indegree[child]--;
                    ancestors.get(child).add(node);

                    for (int kid: ancestors.get(node)) {
                        ancestors.get(child).add(kid);
                    }

                    if (indegree[child] == 0) {
                        queue.add(child);
                    }
                }
            }
        }

        List<List<Integer>> result = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            List<Integer> ancestor = new ArrayList<>(ancestors.get(i));
            Collections.sort(ancestor);
            result.add(ancestor);
        }

        return result;
    }
}
